### PR TITLE
SupraSeal

### PR DESCRIPTION
**Pull request by magik6k**
* [x] FFI interface
* [x] Reasonable build process
  * [x] For now depends on a branch in https://github.com/magik6k/supra_seal/tree/feat/multi-out-paths
* [x] Rebase on main
* [x] Implement batch seal task
* [x] C1
* [x] Slot release in finalize
* [x] Test run
  * [x] Dual-pipeline test run
* [ ] `curio seal batch-checklist` command checking readiness for batch sealing
* [ ] Add disk requirements to batch-cpu calc cmd
* [ ] Make sure the batch sealer doesn't pick up non-cc sectors
* [ ] Auto-detect space for 1/2 pipelines
* [ ] Storage capacity calc

# Build:
```
// Note: requires that the system-wide GCC is version 11.x, or that `gcc-11`/`g++-11` is present locally

make batch

// or calibnet
make batch-calibnet
```

# Usage

1. Update all nodes in the cluster to use this branch (use `make curio` for binaries without batch sealer support - it should be fine to use batch sealer binaries everywhere tho.). Use `make batch` to build batch-capable curio on batch-sealer nodes
2. On batch sealer machine run `curio calc batch-cpu`. This will tell you what batch sizes the installed CPU can support. Note that in case of two-CPU systems performance might suffer significantly due to the CPU interconnect bottleneck. This wasn't tested yet so performance is unknown.
3. Setup a new layer for the batch sealer, call it e.g. `batch-machine1`
  3.1. Set `Subsystems.EnableBatchSeal` to true
  3.2. Set `Seal.LayerNVMEDevices` to a list of nvme devices for layer storage
  3.3. IF your CPU is older, set `Seal.SingleHasherPerThread` to true - this is needed for Zen2 or OLDER (Epyc 7xx2)
4. (optional) Either stop your existing SDR sealers, or set `Subsystems.SealSDRMinTasks` on one of the layers used by single-sector SDR sealers to a value larger than batch size - this is so that the non-batch sealers don't steal work before enough accumulates to start a batch
5. (needed temporarily) Setup NVMe disks dedicated to supraseal layer storage (needed on every boot, future versions of curio will automate this)
```
# make sure you have 36 1G huge pages
sudo sysctl -w vm.nr_hugepages=36

# (in curio repo)
cd extern/supra_seal/deps/spdk-v22.09/   
env NRHUGE=36 ./scripts/setup.sh
```
7. Start the batch-sealer curio node!
8. Attach scratch space for P2 output (sealed sectors + TreeC/TreeR; 32G+36G per sector) (normal filesystem storage, should me NVMe backed; ~500MB/s/gpu + whatever is needed for moving sectors off to long-term storage)
9. Add a batch of CC sectors!
```
curio seal start --now --cc --count 32 --actor f01234 --layers cluster --duration-days 365
```
10. You should see a "Batch.." task running, claimed by the machine, CPU use should look something like the screenshot below
2024-08-02-224523_2495x150_scrot
11. In 3.5 ~ 5h you should see GPU use, this is phase2 computing TreeR / TreeC, should take ~2 mins per sector per GPU (3090-grade)
12. After phase2 is done the rest of pipeline stages should execute normally, after the sector is finalized, the block storage will be released for another batch. A well balanced setup should keep phase1 running at all times, and execute phase2/waitseed/commit in before another batch finishes phase1.

**Storage Recs**:
 
You need 2 sets of nvme drives
* Drives for layers - those need to total 10-20M IOPS, with capacity for the 11 x 32G x batchSize x pipelines - raw unformatted block devices, supraseal uses SPDK which will take them over SPDK is a userspace nvme driver, you won't even see the block device in /dev for those nvmes when it takes them over
* Drives for P2 output - with a filesystem, just something reasonably fast with enough capacity (~70G x batchSize x pipelines). Can be remote storage if you can make it go fast enough (~500MiB/s/GPU)

----
## Random unsorted notes:

gpu fan speed on headless systems
xorg.conf:
```
Section "Device"
    Identifier  "NVIDIA GPU"
    Driver      "nvidia"
    Option      "Coolbits" "4"
EndSection
```

```
#!/bin/bash

# Start Xorg or Xvfb
Xorg -noreset +extension GLX +extension RANDR +extension RENDER -logfile /var/log/Xorg.log -config /etc/X11/xorg.conf :0 &

# Allow some time for Xorg to start
sleep 2

# Set DISPLAY environment variable to use Xorg
export DISPLAY=:0

# Set fan control to manual and set the fan speed
nvidia-settings -a "[gpu:0]/GPUFanControlState=1"
nvidia-settings -a "[fan:0]/GPUTargetFanSpeed=70"

# Keep Xorg running
wait
```

Hugepages must be configured (1g, min 36 pages for 32G sector sealing)
```
sudo sysctl -w vm.nr_hugepages=36

root@:~# cat /proc/meminfo | grep Huge -> should say:
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
FileHugePages:         0 kB
HugePages_Total:      36
HugePages_Free:       18
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:    1048576 kB
Hugetlb:        37748736 kB

```

### CODE REVIEW

File: commit-phase1-output
Comment (magik6k) [inline commit-phase1-output:1]: remove, squash merge will remove this from git tree

File: lib/paths/local.go [hunk 1]
```diff
@@ -999,4 +1026,82 @@ func (st *Local) ReadSnapVanillaProof(ctx context.Context, sr storiface.SectorRe
 	return out, nil
 }
 
+func (st *Local) supraPoRepVanillaProof(src storiface.SectorPaths, sr storiface.SectorRef, sealed, unsealed cid.Cid, ticket abi.SealRandomness, seed abi.InteractiveSealRandomness) ([]byte, error) {
```
Comment (snadrus) [inline lib/paths/local.go:1029]: please describe source of the 'from' and 'to' format for the unit test author.

File: lib/proof/porep_vproof_vanilla.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package proof
```
Comment (snadrus) [inline lib/proof/porep_vproof_vanilla.go:1]: I have no idea what's going on here.

File: lib/proof/sn-comp-sector-seal/main.go [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+	var out [32]byte
+	copy(out[:], in)
+
+	out[31] &= 0b0011_1111
+
+	return out
+}
+
+func main() {
+	outPath := os.Args[1]
```
Comment (snadrus) [inline lib/proof/sn-comp-sector-seal/main.go:61]: A  --help would give me an idea what this does.

File: lib/proof/sn_c1_write_path.txt [hunk 4]
```diff
@@ -0,0 +1,55 @@
+void C1<C>::WriteProofs(const char* filename, bool do_tree, bool do_node) {
```
Comment (snadrus) [inline lib/proof/sn_c1_write_path.txt:1]: rm?

File: lib/supraffi/gosupra/main.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package main
```
Comment (snadrus) [inline lib/supraffi/gosupra/main.go:1]: Change name to 'batch' as post-beta will likely be modified code that Supranational doesn't make.

File: lib/supraffi/seal.go [hunk 6]
```diff
@@ -0,0 +1,303 @@
+//go:build supraseal
```
Comment (snadrus) [inline lib/supraffi/seal.go:1]: And this year's Pulitzer Prize goes to...

File: tasks/sealsupra/supra_config.go [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	if info.CoreCount > 0 {
+		info.ThreadsPerCore = threadCount / info.CoreCount
+	}
+
+	return info, nil
+}
+
+func GenerateSupraSealConfig(info SystemInfo, dualHashers bool, batchSize int, nvmeDevices []string) (SupraSealConfig, error) {
```
Comment (snadrus) [inline tasks/sealsupra/supra_config.go:106]: This rhymes with the CLI command (but could get out of sync in the future). If that's correct, having this be called from there may be better.

File: tasks/sealsupra/supra_config.go [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+	}
+
+	config.Topology.SectorConfigs = append(config.Topology.SectorConfigs, sectorConfig)
+
+	return config, nil
+}
+
+func FormatSupraSealConfig(config SupraSealConfig) string {
+	var sb strings.Builder
+
```
Comment (snadrus) [inline tasks/sealsupra/supra_config.go:222]: w := func(s) { sb.WriteString(s + "\n") }

File: tasks/sealsupra/supra_config.go [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+	for i, device := range config.NVMeDevices {
+		sb.WriteString(fmt.Sprintf("           \"%s\"", device))
+
+		// comma if not last
+		if i < len(config.NVMeDevices)-1 {
+			sb.WriteString(",\n")
+		} else {
+			sb.WriteString("\n")
+		}
+	}
```
Comment (snadrus) [inline tasks/sealsupra/supra_config.go:236]:
```suggestion
w(strings.Join(lo.Map(config.NVMeDevices, func(d, i int) string { return `"`+d.String +`"`}, ",\n"))
```

File: tasks/sealsupra/task_supraseal.go [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+			Gpu: 0,
+			Ram: 16 << 30,
+		},
+		MaxFailures: 4,
+		IAmBored:    passcall.Every(30*time.Second, s.schedule),
+	}
+}
+
+func (s *SupraSeal) Adder(taskFunc harmonytask.AddTaskFunc) {
+}
```
Comment (snadrus) [inline tasks/sealsupra/task_supraseal.go:447]: s.taskFunc = taskFunc ?

File: tasks/sealsupra/task_supraseal.go [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+
+func (s *SupraSeal) schedule(taskFunc harmonytask.AddTaskFunc) error {
```
Comment (snadrus) [inline tasks/sealsupra/task_supraseal.go:449]: when does this get called? what's its contract?

**Review by snadrus — CHANGES_REQUESTED** [review]

**Comment by snadrus** [comment]
Lets document this as Batch Sealing Beta in documentation/en/batchsealing.md

since we hope to 1. modernize deps, 2. get rid of spdk 3. avoid needing Rust bin formats.

File: tasks/sealsupra/task_supraseal.go (hunk 11 above)
Comment (magik6k) [reply tasks/sealsupra/task_supraseal.go:449]: > IAmBored:    passcall.Every(30*time.Second, s.schedule),

File: lib/paths/local.go (hunk 1 above)
Comment (magik6k) [reply lib/paths/local.go:1029]: Which 'from' and 'to'?

File: lib/supraffi/gosupra/main.go (hunk 5 above)
Comment (magik6k) [reply lib/supraffi/gosupra/main.go:1]: Oh, this was just an early tool to test that the mad linker invocation works correctly, will remove

File: tasks/sealsupra/task_supraseal.go (hunk 10 above)
Comment (magik6k) [reply tasks/sealsupra/task_supraseal.go:447]: Nope, this task adds from the bored handler.

File: lib/proof/porep_vproof_vanilla.go (hunk 2 above)
Comment (magik6k) [reply lib/proof/porep_vproof_vanilla.go:1]: Added some comments to files in here

File: tasks/sealsupra/supra_config.go (hunk 9 above)
Comment (snadrus) [reply tasks/sealsupra/supra_config.go]: Many review matters seem unchanged. 
Have you not pushed or are you skipping this?

**Review by snadrus — COMMENTED** [review]
Can you put your PR doc info into documentation/en/batchsealing.md
so that doc is already updated in this PR?

File: tasks/sealsupra/supra_config.go (hunk 9 above)
Comment (magik6k) [reply tasks/sealsupra/supra_config.go]: That was pushed, but missed the loop below

File: lib/proof/sn-comp-sector-seal/main.go (hunk 3 above)
Comment (magik6k) [reply lib/proof/sn-comp-sector-seal/main.go:28]: Added more explicit -h

File: documentation/en/supraseal.md [hunk 12]
```diff
@@ -68,6 +73,13 @@ LayerNVMEDevices = [
   # Add PCIe addresses for all NVMe devices to use
 ]
 
+# Set to your desiced batch size (what the batch-cpu command says your CPU supports AND what you have nvme space for)
```
Comment (snadrus) [inline documentation/en/supraseal.md:76]: Are these build-time or run-time flags?

File: documentation/en/supraseal.md [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+env NRHUGE=36 ./scripts/setup.sh
+```
+
+### PC2 output storage
+
+Attach scratch space storage for PC2 output (batch sealer needs ~70GB per sector in batch - 32GiB for the sealed sector,
+and 36GiB for the cache directory with TreeC/TreeR and aux files)
+
+## Usage
+1. Start the Curio node with the batch sealer layer
```
Comment (snadrus) [inline documentation/en/supraseal.md:136]: example, please

File: documentation/en/supraseal.md [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+2. Add a batch of CC sectors:
+```bash
+curio seal start --now --cc --count 32 --actor f01234 --layers cluster --duration-days 365
+```
+
+3. Monitor progress - you should see a "Batch..." task running
```
Comment (snadrus) [inline documentation/en/supraseal.md:142]: Monitor progress:
With a browser pointed to a curio node running --layer=gui you will see 32 sectors with the same task-id on the po-rep views.

File: documentation/en/supraseal.md [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+4. PC1 will take 3.5-5 hours, followed by PC2 on GPU
+5. After batch completion, the storage will be released for the next batch
+
+## Optimizing
+
+* Balance batch size, CPU cores, and NVMe drives to keep PC1 running constantly
+* Ensure sufficient GPU capacity to complete PC2 before next PC1 batch finishes
+* Monitor CPU, GPU and NVMe utilization to identify bottlenecks
+* Monitor hasher core utilisation
+
```
Comment (snadrus) [inline documentation/en/supraseal.md:152]: * Ensure enough available space on regular drives for 32x (11*32GiB + 32GiB) sectors with layer files to sit 150 epochs, as well as 32 * 32 GiB for the complete sectors to offload to longer-term storage while the next batch arrives on the regular drives.

Comment (snadrus) [reply documentation/en/supraseal.md:152]: this comment fits better below. My number is smaller than yours. Don't we only need 1 batch's layer files at a time, unless more batches complete in the 150 epochs?

**Review by snadrus — CHANGES_REQUESTED** [review]

File: documentation/en/supraseal.md (hunk 12 above)
Comment (snadrus) [reply documentation/en/supraseal.md:76]: Can you give an example of running batch-cpu here?

Comment (snadrus) [reply documentation/en/supraseal.md:76]: should we simply infer these settings instead of asking for them?

File: documentation/en/supraseal.md [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+- At least 256GB RAM, ALL MEMORY CHANNELS POPULATED
+  - Without **all** memory channels populated sealing **performance will suffer drastically**
+- NUMA-Per-Socket (NPS) set to 1
+
+## Setup
+
+### Dependencies
+
+CUDA 12.x is required, 11.x won't work.
+
```
Comment (snadrus) [inline documentation/en/supraseal.md:36]: SPDK?

**Review by snadrus — CHANGES_REQUESTED** [review]

File: documentation/en/supraseal.md (hunk 16 above)
Comment (magik6k) [reply documentation/en/supraseal.md:36]: Part of build, happens when you call `make batch`

File: documentation/en/supraseal.md (hunk 12 above)
Comment (magik6k) [reply documentation/en/supraseal.md:76]: Wouldn't hold this PR up on this, opened an issue for tracking https://github.com/filecoin-project/curio/issues/139

File: documentation/en/supraseal.md (hunk 15 above)
Comment (magik6k) [reply documentation/en/supraseal.md:290]: The storage requirements are already in the "Storage Recommendations" section. I've moved it up before setup for visibility

NI support "just" requires changes to porep challenge reading code, but on the Go side it's almost a no-op, changes are in the later stages of the pipeline.

Synthetic doesn't really make sense with supraseal machines, they just move too fast and have to have 2x space for sectors anyways

File: documentation/en/supraseal.md (hunk 14 above)
Comment (magik6k) [reply documentation/en/supraseal.md]: Added a link to the gui page

**Review by snadrus — APPROVED** [review]

**Pull request merged**